andi/andi.py

```diff
@@ -1,62 +1,318 @@
+    dependency_stack = dependency_stack or []
+    plan_seq = Plan()
+    type_for_arg = {}
+
+    if externally_provided(class_or_func):
+        plan_seq[class_or_func] = {}
+        return plan_seq
+
+    if not is_injectable(class_or_func) and dependency_stack:
+        raise NonProvidableError(
+            "Type {} cannot be provided".format(class_or_func))
 
-    if isinstance(can_provide, Container):
-        can_provide = can_provide.__contains__
+    if class_or_func in dependency_stack:
+        raise CyclicDependencyError(
+            "Cyclic dependency found. Dependency graph: {}".format(
+                " -> ".join(map(str, dependency_stack + [class_or_func]))))
+
+    dependency_stack = dependency_stack + [class_or_func]
+
+    if isinstance(class_or_func, type):
+        cls = cast(Type, class_or_func)
+        arguments = inspect(cls.__init__)
+    else:
+        arguments = inspect(class_or_func)
 
-    result = {}
     for argname, types in arguments.items():
-        for cls in types:
-            if can_provide(cls):
-                result[argname] = cls
-                break
-    return result
+        sel_cls = _select_type(types, is_injectable, externally_provided)
+        if sel_cls is not None:
+            if sel_cls not in plan_seq:
+                plan = _plan(sel_cls,
+                             is_injectable=is_injectable,
+                             externally_provided=externally_provided,
+                             strict=True,
+                             dependency_stack=dependency_stack)
+                plan_seq.update(plan)
+            type_for_arg[argname] = sel_cls
+        else:
+            if strict:
+                if not types:
+                    msg = "Parameter '{}' is lacking annotations in " \
+                          "'{}.__init__()'. Not possible to build a plan".format(
+                        argname, class_or_func)
+                else:
+                    msg = "Any of {} types are required ".format(types)
+                    msg += " for parameter '{}' ".format(argname)
+                    msg += " in '{}.__init__()' but none can be provided".format(
+                        class_or_func)
+                raise NonProvidableError(msg)
+
+    plan_seq[class_or_func] = type_for_arg
+    return plan_seq
+
+
+def _select_type(types, is_injectable, externally_provided):
+    """ Choose the first type that can be provided. None otherwise. """
+    for candidate in types:
+        if is_injectable(candidate) or externally_provided(candidate):
+            return candidate
+
+
+def _ensure_can_provide_func(cont_or_call: Optional[ContainerOrCallableType]
+                             ) -> Callable[[Callable], bool]:
+    if cont_or_call is None:
+        return lambda x: False
+    if isinstance(cont_or_call, Container):
+        return cont_or_call.__contains__
+    return cont_or_call
+
+
+def _ensure_input_type_checks_as_func(can_provide, externally_provided
```

Comment:
As this function is only used once, I think it can be dropped

Reply:
Sure. Dropped here: https://github.com/scrapinghub/andi/pull/8/commits/9d0586c7cef234e32cb7893f0710796901154a77